### PR TITLE
Optimize SUM(x^2), SUM(ABS(X)) and MAX(ABS(X)) to cuDNN reduction.

### PR DESCRIPTION
**Pull request by notoraptor**
@nouiz @abergeron

### CODE REVIEW

**Comment by abergeron** [comment]
I'm wondering why we don't just disable the elemwise fusion.

**Comment by nouiz** [comment]
It will raise the peak memory usage and will be much slower then with this PR.

old code:
- V100: 0.136s
- TITAN Xp: 0.127s

new code:
v100: 0.000235s
Titan Xp: 0.00035s

speed up:
- V100 578x
- Titan Xp 361x

**Comment by abergeron** [comment]
Ok for this then.

**Comment by notoraptor** [comment]
One Jenkins failure: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/961/testReport/junit/theano.gpuarray.tests/test_opt/test_local_gpu_elemwise_careduce/

Fixed: 4f2862a

File: theano/gpuarray/tests/test_opt.py [hunk 1]
```diff
@@ -359,24 +359,19 @@ def test_pdbbreakpoint_op():
...
-    data = np.random.rand(3, 4).astype(theano.config.floatX)
-    utt.assert_allclose(f(data), (data * data).sum())
-
-    o = (x * x).sum(axis=1)
-    f = theano.function([x], o, mode=mode_with_gpu)
-    topo = f.maker.fgraph.toposort()
-    assert len(topo) == 3
-    assert topo[1].op.pre_scalar_op == theano.scalar.sqr
-    assert _check_stack_trace(f)
-    utt.assert_allclose(f(data), (data * data).sum(axis=1))
```
Comment (lamblin) [inline theano/gpuarray/tests/test_opt.py]: Maybe we should keep that test in addition to the new one, but disable dnn optimizations in the mode.
That way we make sure the old optimization still works when there is no cudnn.

**Comment by notoraptor** [comment]
@lamblin Fixed !

File: theano/gpuarray/dnn.py [hunk 2]
```diff
@@ -3746,19 +3746,33 @@ def local_dnn_reduction(node):
...
+        return GpuElemwise(theano.scalar.basic.sqr)(a)
+
+    scal = node.op.scalar_op.name
+    post = _identity
+
+    if (isinstance(node.op.pre_scalar_op, theano.scalar.basic.Sqr) and
+            isinstance(node.op.scalar_op, theano.scalar.basic.Add)):
+        scal = 'norm2'
+        post = _square
+    elif node.op.pre_scalar_op is not None:
```
Comment (nouiz) [inline theano/gpuarray/dnn.py]: While at it, can you add the case for norm1? This way, we shouldn't need to make such fix in the futur. If you can't do that today, I think we should merge this PR, but if it can be done quickly, it can be added to this PR.

**Comment by notoraptor** [comment]
@nouiz updated.

**Comment by notoraptor** [comment]
@nouiz Updated, I added optimization for `MAX(ABS(X))`.

I cannot currently add optimization for average (`MEAN(X)`) because theano `mean()` function produces a graph with a sum divided iteratively by axes (something like `x.sum() / shape[0] / shape[1] / ...`). So I don't know if it is a good idea to parse such case.

Mean function is here: https://github.com/Theano/Theano/blob/master/theano/tensor/basic.py#L3272

File: theano/gpuarray/tests/test_opt.py [hunk 3]
```diff
@@ -372,12 +372,13 @@ def fn_sum_abs(x, axis):
...
-        for fn, pre_scalar_op in ((fn_sum_square, theano.scalar.sqr),
-                                  (fn_sum_abs, theano.scalar.abs_),
-                                  (fn_max_abs, theano.scalar.abs_)):
+    for fn, pre_scalar_op in ((fn_sum_square, theano.scalar.sqr),
+                              (fn_sum_abs, theano.scalar.abs_),
+                              (fn_max_abs, theano.scalar.abs_)):
+        for axis in (None, 0, 1):
             o = fn(x, axis)
             f = theano.function([x], o, mode=mode_with_gpu_no_cudnn)
+            theano.printing.debugprint(f)
```
Comment (nouiz) [inline theano/gpuarray/tests/test_opt.py]: remove from the test

**Comment by nouiz** [comment]
jenkins a eu une erreur

**Comment by notoraptor** [comment]
I am not able to reproduce the problem (tested on: cuda 8 + cudnn 7002/7003, then cuda 9 + cudnn 7003).

**Comment by nouiz** [comment]
jenkins failure: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/968/

jenkins test this

Maybe it was a random error, so I start the test again, just in case.

**Comment by slefrancois** [comment]
jenkins test this

**Comment by slefrancois** [comment]
Jenkins tests were stuck so I killed and restarted manually.

**Comment by notoraptor** [comment]
In fact, error can be reproduced with `floatX=float16`: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/972/console
```
[Theano_GPU]         // rest of function is handled by one warp
[Theano_GPU]         if (threadNum < warpSize) {
[Theano_GPU]             //round up all the partial sums into the first `warpSize` elements
[Theano_GPU]             for (int i = threadNum + warpSize; i < threadCount; i += warpSize)
[Theano_GPU]             {
[Theano_GPU]                 myresult = ((buf[i])>(myresult)? (buf[i]): ((myresult)>=(buf[i])? (myresult): nan("")));
[Theano_GPU]             }
[Theano_GPU]             buf[threadNum] = myresult;
[Theano_GPU]         }
[Theano_GPU]         __syncthreads();
[Theano_GPU]         for (unsigned int _n = warpSize / 2; _n > 0; _n /= 2) {
[Theano_GPU]             if (threadNum < _n && threadNum + _n < threadCount)
[Theano_GPU]             buf[threadNum] = ((buf[threadNum+_n])>(buf[threadNum])? (buf[threadNum+_n]): ((buf[threadNum])>=(buf[threadNum+_n])? (buf[threadNum]): nan("")));
[Theano_GPU]             __syncthreads();
[Theano_GPU]         }
[Theano_GPU]         if (threadNum == 0) {
[Theano_GPU]           Z[0] = (buf[0]);
[Theano_GPU]         }
[Theano_GPU]         
[Theano_GPU]             }
[Theano_GPU]             
[Theano_GPU] NVRTC compile log::
[Theano_GPU] default_program(11): error: more than one instance of overloaded function "fabs" matches the argument list:
[Theano_GPU]             function "fabs(double)"
[Theano_GPU]             function "fabs(float)"
[Theano_GPU]             argument types are: (const short)
```

It seems a call to `fabs()` receives a `short` but can not then choose between `fabs(double)` and `fabs(float)`. It seems related to how the kernel code is generated for pre scalar op. My fix in last commit seems to resolve the problem: https://github.com/Theano/Theano/pull/6491/commits/5e80e5fa46a0e47159c59052ae1e84579b9e155e

@nouiz @abergeron @lamblin

**Comment by nouiz** [comment]
travis fail with wrong result errors.

**Comment by notoraptor** [comment]
Yes, I'm working on it.

**Comment by notoraptor** [comment]
Let's see if jenkins still have errors.

**Comment by notoraptor** [comment]
Still jenkins errors: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/977/

I retry with libgpuarray compiled in `Release` mode.

**Comment by notoraptor** [comment]
Oh, that's strange: no more error when libgpuarray is compiled in Release mode: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/980/

Is it something expectable ? @abergeron @nouiz @lamblin 

I guess no.

**Comment by nouiz** [comment]
I do not expect changing from Debug compilation to Release compilation of libgpuarray to change the results. I'll review in detail this PR. Can you make a new PR that set again the Debug compilation of libgpuarray?

**Comment by notoraptor** [comment]
@nouiz Done: #6496 .

File: theano/gpuarray/elemwise.py [hunk 4]
```diff
@@ -932,7 +932,7 @@ def _k_init(self, node, nodename):
                 Z = (%(out_type)s *)(((char *)Z)+offset_Z);
         """ % locals()
 
-    def _assign_init(self, first_item):
+    def _assign_init(self, first_item, dtype='float32'):
```
Comment (nouiz) [inline theano/gpuarray/elemwise.py]: remove the default value for dtype.

File: theano/gpuarray/elemwise.py [hunk 5]
```diff
@@ -1744,7 +1743,7 @@ def gpu_kernels(self, node, nodename):
             reduce_fct = self._assign_reduce(node, nodename, "myresult",
                                              load_in + "(A[i0])",
                                              {}, True)
-            reduce_init = self._assign_init(load_in + "(A[0])")
+            reduce_init = self._assign_init(load_in + "(A[0])", acc_dtype)
```
Comment (nouiz) [inline theano/gpuarray/elemwise.py]: I think you should use the input dtype, not the acc_dtype in all call to assign_init()

File: theano/gpuarray/opt.py [hunk 6]
```diff
@@ -1207,7 +1207,7 @@ def local_gpua_careduce(op, context_name, inputs, outputs):
             return False
         x, = inputs
         idtype = x.dtype
-        adtype = getattr(op, 'acc_dtype', None)
+        adtype = getattr(op, 'acc_dtype', idtype)
```
Comment (nouiz) [inline theano/gpuarray/opt.py]: Why this change? I don't understand it.

Comment (notoraptor) [reply theano/gpuarray/opt.py]: It's because of this line in optimization that converts gpucareduce to gpudnnreduction: https://github.com/Theano/Theano/blob/master/theano/gpuarray/dnn.py#L3737

The `acc dtype` in CPU careduce may be undefined or set to None, but it is currently updated by the optimization that converts careduce to gpucareducecuda (1) for float32 case, so that everything goes well in float32. But when floatX=float64, `acc dtype` will still be None after optimization from CPU to GPU (1), and then optimization from GPU to cuDNN will fail at line pointed above.

(1): https://github.com/Theano/Theano/blob/master/theano/gpuarray/opt.py#L1210

I detected this problem with `theano.gpuarray.tests.test_dnn.test_dnn_reduction_absmax`, which failed with floatX=float64 on my computer.

With the introduced change, I make sure `acc dtype` falls back to input dtype if it is not defined by CPU careduce.

File: theano/gpuarray/elemwise.py (hunk 4 above)
Comment (notoraptor) [reply theano/gpuarray/elemwise.py]: Fixed

File: theano/gpuarray/elemwise.py (hunk 5 above)
Comment (notoraptor) [reply theano/gpuarray/elemwise.py]: OK

**Comment by notoraptor** [comment]
Updated.

File: theano/gpuarray/opt.py (hunk 6 above)
Comment (nouiz) [reply theano/gpuarray/opt.py]: From tensor/elemwise.py line 1340, I think we should use the output dtype as the acc_dtype when there is no acc_dtype.

Comment (notoraptor) [reply theano/gpuarray/opt.py]: This line (tensor/elemwise.py line 1340) is in `perform()` method, so it does not change the op attributes (and it will not be called when optimizing to GPU).

Comment (notoraptor) [reply theano/gpuarray/opt.py]: PS: Maybe can I just fix it for float64 as it is done for float32 ?  https://github.com/Theano/Theano/blob/master/theano/gpuarray/opt.py#L1216

With float64:
```python
        if idtype == 'float64' and odtype == 'float64':
            adtype = 'float64'
```

Comment (nouiz) [reply theano/gpuarray/opt.py]: This won't be enough. Your change to assing_init are good. We should not always use float32 inside it. I'm still reviewing your new commit.

**Comment by lamblin** [comment]
When libgpuarray is compiled in Debug, it also adds an offset to each memory allocation.
If incorrect results get produced, it could mean that the offset is not correctly passed. Since the offset is zero for most arrays, compiling in Release could work most of the time, but still be incorrect in some cases.

**Comment by nouiz** [comment]
@notoraptor can you try to reproduce those error locally by building libgpuarray in Debug?

if you can reproduce them, try with Theano mode=DebugMode in addtion to the above. This could help identify which op have the problem.

**Comment by notoraptor** [comment]
@nouiz Yes I am able to reproduce it on Ubuntu (by compiling libgpuarray in Debug mode). I will try theano debug mode tomorrow.

Note that DNN errors are fixed by @abergeron 's commit : http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/977/

**Comment by notoraptor** [comment]
(bad jenkins link above, this is the last one in the other PR): http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/984/

**Comment by notoraptor** [comment]
This PR can be closed, as it continues with #6496 .
